core/trino-main/src/main/java/io/trino/operator/PagesIndex.java

```diff
@@ -229,14 +229,14 @@ public void addPage(Page page)
             pagesMemorySize += block.getRetainedSizeInBytes();
         }
 
-        for (int position = 0; position < page.getPositionCount(); position++) {
-            long sliceAddress = encodeSyntheticAddress(pageIndex, position);
+        // this uses a long[] internally, so cap size to a nice round number for safety
+        int resultingSize = valueAddresses.size() + page.getPositionCount();
```

Comment:
Maybe just use a `long` here instead of `resultingSize < 0` for detecting overflow.